src/runtime/internal/scheduler.ts

```diff
@@ -31,17 +31,21 @@ export function add_flush_callback(fn) {
 	flush_callbacks.push(fn);
 }
 
-let flushing = false;
 const seen_callbacks = new Set();
 export function flush() {
-	if (flushing) return;
-	flushing = true;
+
+	let current_component = null;
+	try {
+		current_component = get_current_component();
+	} catch {
+		// no current component, so leave it as null
+	}
 
 	do {
 		// first, call beforeUpdate functions
 		// and update components
-		for (let i = 0; i < dirty_components.length; i += 1) {
-			const component = dirty_components[i];
+		while (dirty_components.length) {
+			const component = dirty_components.shift();
```

Comment:
Is the queue change required for the issues fixed? Otherwise I think we can separate the PR as it's a crucial decision to make. It may be hard to justify the performance gain for the extra bundle size.

Reply:
Is there any way to make this work without the queue? Like tracking the current index outside of the flush function and resetting it when flush is done.

Reply:
Thanks for the idea, @dummdidumm. That worked, so I've implemented it (and removed the Queue implementation) in https://github.com/sveltejs/svelte/pull/6920/commits/83e9cc5d5d405c9cbd20176c0bc636bb782178a6. Note that I put the index increment on its own line instead of doing `dirty_components[flushidx++]`. I've never liked the `array[idx++]` and `array[++x]` idioms; I view them as developer traps, mistakes waiting to happen. Any time it's possible to be more explicit about whether the increment is happening before or after the array deference, I prefer to be explicit.

But if you (plural you, the core Svelte team) want this to be written as `dirty_components[flushidx++]`, I'm willing to make that change.